internal/rest/leveldbreceipts.go

```diff
@@ -0,0 +1,373 @@
+func (l *levelDBReceipts) AddReceipt(requestID string, receipt *map[string]interface{}) (err error) {
+	// insert an entry with a composite key to track the insertion order
+	newId := ulid.MustNew(ulid.Timestamp(time.Now()), l.idEntropy)
+	// add "z" prefix so these entries come after the lookup entries
+	// because for iteration we start from last backwards
+	lookupKey := fmt.Sprintf("z%s", newId)
+
+	b, _ := json.MarshalIndent(receipt, "", "  ")
+	l.store.Put(lookupKey, b)
+
+	// build the index for "from"
+	fromKey := fmt.Sprintf("from:%s:%s", (*receipt)["from"], lookupKey)
+	l.store.Put(fromKey, []byte(lookupKey))
+
+	// build the index for "to" if a value is present
+	to := (*receipt)["to"]
+	if to != "" {
+		toKey := fmt.Sprintf("to:%s:%s", to, lookupKey)
+		l.store.Put(toKey, []byte(lookupKey))
+	}
+
+	// build the indiex for "receivedAt"
+	receivedAtKey := fmt.Sprintf("receivedAt:%d:%s", (*receipt)["receivedAt"], lookupKey)
+	l.store.Put(receivedAtKey, []byte(lookupKey))
+
+	// insert the lookup entry for GetReceipt()
+	return l.store.Put(requestID, []byte(lookupKey))
+}
+
+// GetReceipts Returns recent receipts with skip, limit and other query parameters
+func (l *levelDBReceipts) GetReceipts(skip, limit int, ids []string, sinceEpochMS int64, from, to, start string) (*[]map[string]interface{}, error) {
+	// the application of the parameters are implemented to match mongo queries:
+	// - find the starting point:
+	//   - if "start" is present, use it
+	//   - otherwise use "Last()"
+	// - if "skip" is present, forward to the count
+	// - if "ids" are present, use them to look up the specific entries and filter out the entries falling out of the cursor range
+	// - if "from" or "to" are present, look up the entries using the "from:[address]" and "to:[address]" prefix then work out the intersection of the [lookupKey] segments
+	var itr kvstore.KVIterator
+	var endKey string
+	results := []map[string]interface{}{}
+	if sinceEpochMS > 0 {
+		// locate the iterator range limit
+		keyRange := l.findEndPoint(sinceEpochMS)
+		if keyRange != nil {
+			itr = l.store.NewIterator(keyRange)
+			endKey = string(string(keyRange.Limit))
+		} else {
+			// no entries match the sinceEpochMS, return empty
+			return &results, nil
+		}
+	} else {
+		// create the iterator without a range
+		itr = l.store.NewIterator()
+	}
+	defer itr.Release()
+
+	if limit == 0 {
+		limit = l.defaultLimit
+	}
+
+	// if we have reference points, like list of IDs and "from/to" parameters, process using the reference points
+	var lookupKeys []string
+	var lookupKeysByIDs []string
+	if len(ids) > 0 {
+		// use the list of IDs to search for the result entries
+		lookupKeysByIDs = l.getLookupKeysByIDs(ids, start, endKey)
+	}
+	var lookupKeysByFromAndTo []string
+	if from != "" || to != "" {
+		lookupKeysByFromAndTo = l.getLookupKeysByFromAndTo(from, to, start, endKey, limit)
+	}
+	if len(ids) > 0 && from == "" && to == "" {
+		lookupKeys = lookupKeysByIDs
```

Comment:
This path doesn't seem to drop into the `results := l.getReceiptsByLookupKey(lookupKeys, limit)` branch (because of an `else if`) - so this result is un-used.

Reply:
all 3 branches should be used:
- if `ids` are present, and both `from` and `to` are empty, branch 1
- if `ids` are empty, and at least one of `from` and `to` is non-empty, branch 2
- if `ids` are present, and one of `from` and `to` is non-empty, branch 3

Reply:
Agreed - I just mean in the code, there are two branches where all they do is set `lookupKeys`, but there's no reference to `lookupKeys` below that in the function.

Reply:
I added a commit and unit test to resolve this.